### PR TITLE
feat: add `ArgillaSpaCyTrainer` for both `TokenClassification` and `TextClassification`

### LINKED ISSUE #2594: Add `argilla.training` module for spaCy integration
Currently, we are working on support for `transformers` and already wrapped up support for `setfit`. Ideally, we want to include support for `spacy` too. By default, we would like to see support for `token_classificaiton`, but due to the `prepare_for_training(framework="spacy", nlp=nlp)` method within Argilla, it should be straightforward to add `text_classification` support too. You can just look at the setfit implementation, but the outcome of the PR should be.

```python
from argilla.training import ArgillaTrainer
​
trainer = ArgillaTrainer("dataset_name", "model_path_or_name" framework="spacy", train_size=0.8)
trainer.update_config(num_epochs=2, **kwargs)
trainer.train(path="test_model")
records = trainer.predict("This is a test", as_argilla_records=True)
```

Even though spaCy focuses on CLI-based training it should be accessible via the Python client too and can thus be nicely integrated here too.

```python
from spacy.cli.train import train as spacy_train
config_path = "spacy_textcat/config.cfg"
output_model_path = "output/spacy_textcat"
spacy_train(
    config_path,
    output_path=output_model_path,
    overrides={
        "paths.train": "train.spacy",
        "paths.dev": "valid.spacy",
    },
)
```

Including some hyperparameter optimization might also be an option.

_Originally posted by @davidberenstein1957 in https://github.com/argilla-io/argilla/issues/2504#issuecomment-1480981433_

### PR DESCRIPTION
**Pull request by alvarobartt**
# Description

This PR adds `ArgillaSpaCyTrainer` to use `spaCy` to train both `token-classification` and `text-classification` models through as of the recent development of `argilla.training`. `ArgillaSpaCyTrainer` leverages the features of [`spacy train`](https://spacy.io/usage/training) programmatically, to simplify the integration with Argilla.

This PR has been created on top of https://github.com/argilla-io/argilla/pull/2574 which is the former PR created by @davidberenstein1957 to add `argilla.training`.

Closes #2594

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

No unit tests have been developed as part of this PR since those will either be pushed as part of the main PR at https://github.com/argilla-io/argilla/pull/2574, or on another PR before merging into `develop`.

I took the chance to create a Jupyter Notebook following the `template.ipynb` provided by Argilla. The Jupyter Notebook shows how to train a NER from scratch using `spaCy`, and is available at https://colab.research.google.com/drive/1zwIpSMMhfyuUmuy2lMpp-HbZc-un8S2N?usp=share_link

**Checklist**

- [X] I have merged the original branch into my forked branch
- [x] I added relevant documentation
- [X] follows the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

### CODE REVIEW

**Comment by alvarobartt** [comment]
As of the latest commit, the following script is working successfully:

```python
import argilla as rg
from argilla.training import ArgillaTrainer

rg.init(api_url="https://alvarobartt-argilla-datasets.hf.space", api_key="team.apikey")

trainer = ArgillaTrainer(name="ner", model="en_core_web_trf", framework="spacy", train_size=0.8)
trainer.update_config(max_epochs=10)
trainer.train(path="ner")
records = trainer.predict("I live in Barcelona.", as_argilla_records=True)

rg.log(
    records=records,
    name="ner",
    tags={
        "task": "NER",
        "family": "token-classification",
        "dataset": "gutenberg-time",
    },
)
```

**Comment by alvarobartt** [comment]
I'd like someone's input on how's the best way to define the configuration, since `spaCy` expects either a nested Python dict which breaks with the current approach of providing params via named kwargs, or a file (which is easier to do, but requires more effort from the user).

So on, to be aligned with `ArgillaTransformersTrainer` and `ArgillaSetfitTrainer`, I've decided to add just the `training` params to be "tweakable" by using `update_config`.

This may not be the best approach in case the user wants either to change other nested values e.g. `training->batcher->tolerance`, as now just the following parameters can be tweaked from `update_config`:

```ini
[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
annotating_components = []
before_to_disk = null
before_update = null
```

So I'd like to clarify this or brainstorm some ideas before I advance deeper on configuration-related stuff :hugs:

Thanks in advance!

cc @frascuchon @davidberenstein1957 @tomaarsen

**Comment by davidberenstein1957** [comment]
Hi, @alvarobartt, great to see the progress and good that you ask a question. For me, a default initialization of the config could be good enough for an MVP.

However, given that you can define custom `update_config`-behavior, I would like you to see if you can create a method that can do an iterative deep update of the nested dict variables after the default initialization. We normally work with pydantic models to define these data-structures, so that might be of some help and guidance for the user too.

Given the time indication, it might be more interesting in first looking into creating an MVP for `token-classification` and `text-classification` with the default variables, while just looking for the 

Good luck.

**Comment by alvarobartt** [comment]
Hi @davidberenstein1957, we can keep the default initialization for now! Anyway, I'll try to come up with a more detailed solution 👍🏻 

I'll be AFK during most of the day today, but hopefully, I'll be able to push the remaining stuff by the end of the day

Thanks for the suggestion :hugs:

**Comment by tomaarsen** [comment]
Hey Alvaro! Good to see you here as well. I'm looking forward to seeing how this PR shapes up.

**Comment by alvarobartt** [comment]
Hi @davidberenstein1957! I also wanted to comment on another topic related to `spaCy` training, which is that it expects the data to be dumped in a file before training, so we need to provide a path, we cannot provide a buffer. What is the best way to handle this? I was thinking of relying on a cache where the name of the subdir could be a hash generated out of the dataset's full URL, but since the idea here is to create an annotation-training loop, maybe that's not the most functional stuff... Since this is an important matter I'd like to get your input first 👍🏻 I'm currently relying on the cache, but I also see potential issues there...

**Comment by alvarobartt** [comment]
As of the latest commit, both `textcat` and `textcat_multilabel` pipelines can be trained successfully with `ArgillaSpaCyTrainer`!

* `textcat`

```python
import argilla as rg
from argilla.training import ArgillaTrainer

rg.init(api_url="https://alvarobartt-argilla-datasets.hf.space", api_key="team.apikey")

trainer = ArgillaTrainer(name="textcat", framework="spacy", train_size=0.8)
trainer.update_config(max_epochs=2)
trainer.train(path="textcat")
records = trainer.predict("I loved the last Transformers movie!", as_argilla_records=True)
rg.log(
    records=records,
    name="textcat",
)
```

* `textcat_multilabel`

```python
import argilla as rg
from argilla.training import ArgillaTrainer

rg.init(api_url="https://alvarobartt-argilla-datasets.hf.space", api_key="team.apikey")

trainer = ArgillaTrainer(name="textcat_multilabel", framework="spacy", train_size=0.8)
trainer.update_config(max_epochs=2)
trainer.train(path="textcat_multilabel")
records = trainer.predict("I'm really grateful for this opportunity!", as_argilla_records=True)
rg.log(
    records=records,
    name="textcat_multilabel",
)
```

P.S. I used `rotten_tomatoes` and `go_emotions` datasets from the HF Hub for the experiments above, respectively. For the sake of testing I set the labels as annotations in Argilla for fast experimentation without having to manually annotate the examples. Anyway, the datasets I'm hosting in my HF Space for Argilla may not be consistent, so eventually the experiments above may not work with those `api_url` and `name`.

**Comment by alvarobartt** [comment]
So the main concerns before moving this PR out of draft are:

* Data needs to be dumped in local storage to be used with `spaCy` training loop
* Just the training configuration can be updated via `update_config`, due to the global configuration being a nested Python dict (we could use `BaseModel` to easily solve that, but that may break with the ease of usage that we currently have)
* Missing unit tests as not sure whether those are WIP in the main PR (LMK if you want me to work on the `argilla.training.base` and/or `argilla.training.spacy`
* No documentation pages have been created/edited as I assume we need to wait upon https://github.com/argilla-io/argilla/pull/2591 completion
* `CHANGELOG.md` hasn't been updated yet as it should be updated once this branch is merged into #2574 

Thanks in advance! cc @davidberenstein1957 @tomaarsen @frascuchon

**Comment by alvarobartt** [comment]
In the mean time I also prepared a simple Jupyter Notebook tutorial in Google Colab with an example on how to train a NER from scratch using `ArgillaSpaCyTrainer`! Available at https://colab.research.google.com/drive/1zwIpSMMhfyuUmuy2lMpp-HbZc-un8S2N?usp=share_link

**Comment by davidberenstein1957** [comment]
Thanks for the PR. It looks great.

> So the main concerns before moving this PR out of draft are:
> 
>     * Data needs to be dumped in local storage to be used with `spaCy` training loop

This is fine if there is no other way to do this.

>     * Just the training configuration can be updated via `update_config`, due to the global configuration being a nested Python dict (we could use `BaseModel` to easily solve that, but that may break with the ease of usage that we currently have)

Perhaps we can add a flag/highlight config options to choose whether to only update the NER/textcat and leave other components intact? This can potentially be set by default and be configured or mentioned by updating the `trainer.__repr__` accordingly. My idea behind the `argilla.training` was to the 80-20 paradigm, assuming that most people just want some basic config and that the people who do want total customization will setup things up themselves.

>     * Missing unit tests as not sure whether those are WIP in the main PR (LMK if you want me to work on the `argilla.training.base` and/or `argilla.training.spacy`

These will be added along with the bas trainer and other components. For your application, you don't need to work on it given the time restriction, however, as an open-source enthusiast, these PRs are always welcome.

>     * No documentation pages have been created/edited as I assume we need to wait upon [feat: Train: Adding code snippets #2591](https://github.com/argilla-io/argilla/pull/2591) completion

Correct.

>     * `CHANGELOG.md` hasn't been updated yet as it should be updated once this branch is merged into [Feat/2504 add an argillatraining module to the core library with training workflows #2574](https://github.com/argilla-io/argilla/pull/2574)

Correct.

> Thanks in advance! cc @davidberenstein1957 @tomaarsen @frascuchon

**Comment by alvarobartt** [comment]
Thanks for the comments @davidberenstein1957 🎉 I already included a disclaimer in the `update_config` docstrings, but I can do the same in the `__repr__` function so that the user knows that just the "training" config can be updated!

**Comment by alvarobartt** [comment]
With the latest change on `__repr__`, now the output in the CLI looks like:

```
The parameters are configurable via `trainer.update_config()`:                                                           
    WARNING:`ArgillaSpaCyTrainer.update_config` just supports the update of the `training` args defined in the `config.yaml` file for the sake of simplicity.                                                                       
                                                                                                                                                    
    `ArgillaSpaCyTrainer`                                                                                                
        dev_corpus: corpora.dev                                                                                      
        train_corpus: corpora.train                                                                                  
        seed: ${system.seed}                                                                                         
        gpu_allocator: ${system.gpu_allocator}                                                                       
        dropout: 0.1                                                                                                 
        accumulate_gradient: 1                                                                                       
        patience: 1600                                                                                               
        max_epochs: 0                                                                                                
        max_steps: 20000                                                                                             
        eval_frequency: 200                                                                                          
        frozen_components: []                                                                                        
        annotating_components: []                                                                                    
        before_to_disk: None                                                                                         
        before_update: None                                                                                          
```
